File: tests/python/test_dropout.py
import sys, os
here = os.path.dirname(__file__)
root = os.path.normpath(os.path.join(here, '..', '..', 'python'))
sys.path.insert(0, root)
from tinyfin import Tensor
import importlib.util
nn_path = os.path.join(root, 'nn.py')
spec = importlib.util.spec_from_file_location('tinyfin_nn', nn_path)
nn = importlib.util.module_from_spec(spec)
spec.loader.exec_module(nn)

Dropout = nn.Dropout

# simple smoke test
x = Tensor.new([2,3])
arr = x.to_numpy()
arr[:] = [[1.0, 1.0, 1.0],[1.0,1.0,1.0]]
D = Dropout(p=0.5)
D.train()
y = D(x)
# during training, some elements should be zeroed (statistically)
yn = y.to_numpy()
assert yn.shape == (2,3)
D.eval()
y2 = D(x)
# during eval, output equals input
y2n = y2.to_numpy()
assert (y2n == x.to_numpy()).all()
print('[test_dropout.py] PASS')

File: tests/test_dropout.c
#include "tensor.h"
#include "ops_dropout.h"
#include <stdio.h>
#include <math.h>

int main() {
    int shape[2] = {2,4};
    Tensor *x = tensor_new(2, shape);
    for (int i=0;i<8;i++) x->data[i] = 1.0f;

    Tensor *y = tensor_dropout(x, 0.5f, 1);
    if (!y) { printf("dropout returned NULL\n"); return 1; }
    int zeros = 0;
    for (int i=0;i<8;i++) if (y->data[i] == 0.0f) zeros++;
    if (zeros == 0) { printf("expected some zeros during training\n"); return 2; }

    tensor_free(y);
    Tensor *z = tensor_dropout(x, 0.5f, 0);
    for (int i=0;i<8;i++) if (fabsf(z->data[i] - x->data[i]) > 1e-6f) { printf("eval path differs\n"); return 3; }

    tensor_free(z); tensor_free(x);
    printf("[test_dropout.c] PASS\n");
    return 0;
}
